### PR TITLE
Add Plotly-like interpolation algorithm to `optuna.visualization.matplotlib.plot_contour`

### LINKED ISSUE #2738: Make `optuna.visualization.matplotlib.plot_contour` more beautiful
## Motivation

`plot_contour` with Matplotlib backend looks different from one with Plotly backend. 

## Description

As shown in https://github.com/optuna/optuna/pull/2712#pullrequestreview-679629017, we can improve a plot generated by using `optuna.visualization.matplotlib.plot_contour` by following points:

- [ ] smoother contour lines
- [ ] fill blank area

### PR DESCRIPTION
**Pull request by xadrianzetx**
This PR introduces new interpolation algorithm to be used when creating contour plots with `optuna.visualization.matplotlib.plot_contour()`, improving on linear interpolation offered by `scipy.griddata()`. 

The algorithm is a port of [Plotly.js](https://github.com/plotly/plotly.js/blob/master/src/traces/heatmap/interp2d.js#L30) version.

This addresses known issues in current implementation:

* contour lines which are not smooth
* blank areas on edges of the plot

Additionally, tweaks to `xrange` and `yrange`  are introduced in order to correctly visualize trial values at the edges of param grid.

## Motivation
This PR resolves #2738 and addresses [#2712 (review)](https://github.com/optuna/optuna/pull/2712#pullrequestreview-679629017).

## Description of the changes
- [x] Implement interpolation algorithm and helpers.
- [x] Switch to new algorithm in `optuna.visualization.matplotlib.plot_contour()`
- [x] Address `xrange` and `yrange` padding
- [x] Write tests

## Examples
All plots were generated using example from #2595 

Current algorithm
image

New algorithm
image

Plotly
image

And a very cool gif showing how this algorithm converges :^)
plot_iter2

### CODE REVIEW

**Comment by xadrianzetx** [comment]
`blackdoc` itself has failed in CI (not the files it was checking). Strange.

**Comment by xadrianzetx** [comment]
Waiting for #2806 to merge so I can proceed with tests.

**Comment by hvy** [comment]
A friendly ping that #2806 has been merged 🙂

**Comment by hvy** [comment]
Assigned @himkt although the work is still ongoing.

And, thanks a lot for the great work! At a quick glance, looks like a definite improvement. I'm just curious how critical Numba is to this functionality. Does the visualization become unusable without the JIT compilation (i.e. how much faster does it become with Numba)? We want to be careful introducing additional install requirements.

**Comment by xadrianzetx** [comment]
@hvy thank you for the question! Let me share a small bench test i did while developing this.

| Version  | Runtime|
| ------------- | ------------- |
| Matplotlib (no JIT) | 1 loop, best of 5: 1min 17s per loop |
| Matplotlib (JIT) | 1 loop, best of 5: 880 ms per loop |
| Plotly | 1 loop, best of 5: 304 ms per loop  |

I understand your point about introducing additional install requirements though. As an alternative, i can move jitted code to separate module, `try_import` it, and fallback to default `griddata` solution if we are missing dependencies. This would move `numba` out of install requirements back to optional. Then i could add a note in dosctring of `plot_contour` informing users about this functionality. Please let me know what do you think of this solution.

**Comment by hvy** [comment]
Thanks for the numbers! With a separate module, do you mean independent of Optuna or within the Optuna library?

Interesting that Plotly is significantly faster than the Python implementation without the JIT compilation. I haven't taken a deeper look at the algorithm yet, but I'm curious whether there is room for optimizing the Python implementation without relying on Numba (e.g. by simple vectorization). If there is, we could perhaps leave it as future work and introduce this algorithm as the default one (given that the matplotlib visualizations are all still experimental).

**Comment by hvy** [comment]
@nzw0301 let me assign you as the author of the issue. Let me know if you're busy and I can reassign someone else. Do you have any thoughts or inputs on how to continue forward with this?

**Comment by xadrianzetx** [comment]
@hvy, sorry for not being very clear on that one. I meant module within Optuna, more specifically within `optuna.visualization.matplotlib` with imports working in the similar way to `_matplotlib_imports.py`.

I think Plotly speed comes from the fact that there is not much heavy lifting done python side. From what i know, execution is passed almost immediately to one of [renderers](https://plotly.com/python/renderers/).

I'm not sure if vectorization is feasible in this instance, since algorithm relies on very specific order in which missing values are interpolated (i. e. when imputing `v[1]` with mean of neighbors, there might be a case when its necessary to fill `v[0]` first, when those values are close to each other on the grid). There are two way I think this algorithm could be speed up if we were to drop Numba:

* Interpolate z-matrix first, upsample to smooth contours later (currently its done other way around) - this would mean we are iterating over smaller matrices.
* Instead extensively using array indexing/slicing, base this algorithm on hashmaps of coordinates, and iterate those.

I will explore both possibilities, before marking this as ready to review.

**Comment by xadrianzetx** [comment]
@hvy, so I've managed to cut no-JIT time from 1m 17s to 4.5s (on Intel Core i5) just by using hashmaps. This comes with no  downsides to the quality of the plot and does not use any new dependencies. It's still a bit slower than Plotly or jitted numpy, but I think we are within reasonable runtime for a function that produces a plot. I don't think there is much to gain anymore with pure python or without diverging from Plotly algorithm. Overall im happy to exclude Numba and switch to the new approach, so please let me know if this is the level of performance we are ok with.

**Comment by hvy** [comment]
That sounds great! 
Could you push the change (or create a separate PR if you like), and we could take a look at the implementation? I'm a bit curious about the time complexity how it scales, but it in any case sounds like a significant improvement worth continuing with.

**Comment by nzw0301** [comment]
hi, @xadrianzetx, sorry for the delayed response. We have had national since last Thursday. 

I'll take a look your this PR soon. Please give me a little bit more time.


best regards,

**Review by nzw0301 — CHANGES_REQUESTED** [review]
Thank you for your great PR! The separated functions are really helpful to follow the logic and tests! I do not still understand the interpolation logic yet, but let me leave minor comments :)

**Comment by xadrianzetx** [comment]
Thank you for the comments @nzw0301! Just FYI, it might be couple days before I'm able to address them, due to some hardware failures I'm currently resolving. Sorry for any delays!

**Comment by nzw0301** [comment]
No worries! It is totally fine. I hope your hardware issue is resolved🤞🏻

**Comment by nzw0301** [comment]
hi, let me give additional comments. 

Personally using complex numbers looks cool! However, it is slightly hard for other committers to maintain the codes. So could you replace complex numbers with tuples or `np.ndarray` of two integers if the implementation does not make the computing performance severely slower? Related to this, `zmap` can be replaced with `np.ndarray` whose values are initialised at `np.nan` like `np.full((len(xi), len(yi)), np.NaN)` in `_create_zmap`

In this complex number-free implementation, one drawback is that we need to care the boundary access with `NEIGHBOR_OFFSETS` cared for in `_find_coordinates_where_empty` and `_run_iteration`, like `np.maximum(np.minimum(coord + offset, contour_point_num-1), 0)`.

What do you think about it?

File: optuna/visualization/matplotlib/_contour.py
Comment (himkt) [reply optuna/visualization/matplotlib/_contour.py:445]: I agree with @nzw0301. And after changing the type, we can remove `type: ignore` in _run_iteration.
https://github.com/optuna/optuna/pull/2810/files#diff-72e0c8aff9262dd0913297924afecfa21e461e8628673c43a748e6f809c5cc6dR575

File: tests/visualization_tests/matplotlib_tests/test_contour.py
Comment (himkt) [inline tests/visualization_tests/matplotlib_tests/test_contour.py:41]:
```suggestion
    zmap = {0 + 0j: 1.0, 1 + 0j: 2.0, 0 + 1j: 3.0, 1 + 1j: 4.0}
```

Type fixing.

File: tests/visualization_tests/matplotlib_tests/test_contour.py
Comment (himkt) [inline tests/visualization_tests/matplotlib_tests/test_contour.py:43]:
```suggestion
    zmatrix = _create_zmatrix_from_zmap(zmap, contour_point_num)
```

File: tests/visualization_tests/matplotlib_tests/test_contour.py
Comment (himkt) [inline tests/visualization_tests/matplotlib_tests/test_contour.py:52]:
```suggestion
    zmap = {0 + 0j: 1.0, 1 + 1j: 4.0}
```
Type fixing. ☝️

File: tests/visualization_tests/matplotlib_tests/test_contour.py
Comment (himkt) [inline tests/visualization_tests/matplotlib_tests/test_contour.py:54]:
```suggestion
    empties = _find_coordinates_where_empty(zmap, contour_point_num)
```

File: tests/visualization_tests/matplotlib_tests/test_contour.py
Comment (himkt) [inline tests/visualization_tests/matplotlib_tests/test_contour.py:66]:
```suggestion
    zmap = {0 + 0j: 1.0, 1 + 1j: 4.0}
```

Type fixing.

File: tests/visualization_tests/matplotlib_tests/test_contour.py
Comment (himkt) [inline tests/visualization_tests/matplotlib_tests/test_contour.py:70]:
```suggestion
    zmap, max_fractional_change = _run_iteration(zmap, empties)
```

Sorry for the multiple suggestion. I remove a line suppressing mypy.

File: tests/visualization_tests/matplotlib_tests/test_contour.py
Comment (himkt) [inline tests/visualization_tests/matplotlib_tests/test_contour.py:82]:
```suggestion
    zmap = {0 + 0j: 1.0, 1 + 1j: 4.0}
```

Type fixing!

File: tests/visualization_tests/matplotlib_tests/test_contour.py
Comment (himkt) [inline tests/visualization_tests/matplotlib_tests/test_contour.py:85]:
```suggestion
    zmap = _interpolate_zmap(zmap, contour_point_num)
```

**Review by himkt — CHANGES_REQUESTED** [review]
Huge thanks for the awesome improvements @xadrianzetx!
Let me leave some suggestions about type checking.

**Comment by xadrianzetx** [comment]
@nzw0301,

> zmap can be replaced with np.ndarray whose values are initialised at np.nan like np.full((len(xi), len(yi)), np.NaN) in _create_zmap.

That was the design of initial version introduced in a27cdf4011bd0b7006dac6f8a6d81ac903c52c74 and f7b343f8f46ad0a684020e6d3649778ebf6634d8. Unfortunately, without JIT, performance [really suffered](https://github.com/optuna/optuna/pull/2810#issuecomment-882512050) due to iterative nature of this algorithm, and Numpy element access being [somewhat slow](https://stackoverflow.com/a/22239281). Without Numba, we needed a data structure that offers fast access to its elements, preferably included in Python standard library, hence usage of hashmap.

> So could you replace complex numbers with tuples or np.ndarray of two integers if the implementation does not make the computing performance severely worse?

The reason behind coordinates being expressed as complex numbers is, that it offers the fastest way to calculate locations of neighbors, with just one simple `add` to cover both `x offset` and `y offset` simultaneously. This is one of the most critical parts of the algorithm, which, assuming 200 Optuna trials, gets hit `9800 missing values * 100 optim iterations * 4 neighbors` (excluding nan discovery part).  I did a quick benchmark for both tuples and Numpy arrays (just the core part of addition operation looped over 3 milion iterations) and saw:

* tuple of coordinates (defined as `tuple(sum(c) for c in zip(coord, offset))`) being ~2x slower than `complex`
* Numpy arrays (defined as `coord_arr + offset_arr`) being ~1.4x slower than `complex`

Given those, I would be in favor of staying with current implementation. I do agree with you on maintainability point though, this algorithm is not the easiest thing to understand. To help with this, I think additional comments explaining grid system and algorithm steps could be introduced.

File: optuna/visualization/matplotlib/_contour.py
Comment (xadrianzetx) [reply optuna/visualization/matplotlib/_contour.py:445]: Nice catch, thank you!

**Comment by xadrianzetx** [comment]
Thanks for reviews @nzw0301, @himkt! I've addressed all points in 34a887e640064bcacde35b1cd20eb13da419cb71, and added some additional comments in b83041040077f7cfa1a72875a7b2017cc3627eac.

**Comment by himkt** [comment]
@xadrianzetx 

Thank you so much for the revise! Basically LGTM. ⭐ 
I have one question.

There still be little bit difference between plotly's plot and matplotlib's plot.
Do you know the cause of this difference?

Plotly | matplotlib
 ----   |     ---
スクリーンショット 2021-07-30 001921 | スクリーンショット 2021-07-30 001935

- matplotlib: https://67626-122299416-gh.circle-artifacts.com/0/docs/build/html/reference/visualization/generated/optuna.visualization.matplotlib.plot_contour.html#optuna.visualization.matplotlib.plot_contour
- plotly: https://67626-122299416-gh.circle-artifacts.com/0/docs/build/html/reference/visualization/generated/optuna.visualization.plot_contour.html#optuna.visualization.plot_contour

**Comment by xadrianzetx** [comment]
@himkt,

It might be due to the fact that for Plotly contour we are not doing any upsampling to trial params, so in your case matrix of shape 30x30 (roughly :)) is interpolated and contour smoothing is used, while in matplotlib implementation we are always resampling to 100x100 either with linspace or logspace. This could result in differences we see, especially for small number of trials. 

Other than that, there might be slight differences in implementation between Plotly and matplotlib as far as calculating number and position of contours (looking at your example, matplotlib seems to be more detailed even though we are not specifying number of contour lines to use anywhere). To my best knowledge they both use [marching squares](https://en.wikipedia.org/wiki/Marching_squares), but I did not look up specific implementations.

**Review by nzw0301 — APPROVED** [review]
Thank you for your comments and update PR! I understand the reason for using complex numbers, so I'm happy with the current implementation!

**Comment by himkt** [comment]
> It might be due to the fact that for Plotly contour we are not doing any upsampling to trial params, so in your case matrix of shape 30x30 (roughly :)) is interpolated and contour smoothing is used, while in matplotlib implementation we are always resampling to 100x100 either with linspace or logspace.

It makes sense. I agree with you that it results the difference. And I also think it is not bad point. :+1:

**Review by himkt — APPROVED** [review]
Again, thank you so much @xadrianzetx for the awesome improvement! LGTM.

**Pull request merged**
